### PR TITLE
Refactor binary expression parser to reduce call depth

### PR DESCRIPTION
**Pull request by arv**
Instead of having to call all of the following:

parseLogicalOR_
parseLogicalAND_
parseBitwiseOR_
parseBitwiseXOR_
parseBitwiseAND_
parseEquality_
parseRelational_
parseShiftExpression_
parseAdditiveExpression_
parseMultiplicativeExpression_
parseExponentiationExpression_
parseUnaryExpression_

We just call parseUnaryExpression_ directly and check if there is
an operator after it. We then use precedence to determine the order
to create the binary expression AST node in.

This gives 10% - 20% performance improvement on the esprima perf
test.

### CODE REVIEW

**Comment by arv** [comment]
@johnjbarton PTAL

File: src/syntax/Parser.js [hunk 1]
```diff
@@ -2878,271 +2878,81 @@ export class Parser {
...
     let start = this.getTreeStartLocation_();
-    let left = this.parseMultiplicativeExpression_();
-    while (this.peekAdditiveOperator_(this.peekType_())) {
-      let operator = this.nextToken_();
-      let right = this.parseMultiplicativeExpression_();
-      left = this.newBinaryExpression_(start, left, operator, right);
-    }
-    return left;
+    let left = this.parseUnaryExpression_();
+    return this.parseBinaryExpressionHelper_(start, left, -1, expressionIn);;
```
Comment (johnjbarton) [inline src/syntax/Parser.js]: `;;`

Comment (arv) [inline src/syntax/Parser.js]: Done.

**Comment by arv** [comment]
Travis timeout is bad news on this one too.

File: src/syntax/Parser.js [hunk 2]
```diff
@@ -2715,7 +2715,7 @@ export class Parser {
    *
    * @return {ParseTree}
    */
-  parseExpression_(expressionIn = Expression.IN) {
+  parseExpression_(expressionIn = Expression.NORMAL) {
```
Comment (johnjbarton) [inline src/syntax/Parser.js]: I realize this was a drive-by fix, but `Expression` is very overloaded in this file. Maybe the const type could be `InVariant` or `InExpressionVariant`.

**Comment by johnjbarton** [comment]
LGTM

**Comment by johnjbarton** [comment]
Yep, I think V8 has an issue.  When I have time I'd like to try different node version with the head of that branch with the timing output.

File: src/syntax/Parser.js (hunk 2 above)
Comment (arv) [inline src/syntax/Parser.js]: I have a follow up task to replace this with a simple boolean parameter (with a constant for readability)
